### PR TITLE
[Improvement](functions)Optimized some datetime function's return value

### PR DESCRIPTION
**Pull request by zclllyybb**
# Proposed changes

## Problem summary

make function `year` `month` `day` `hour` `minute` `second` `quarter` `dayofweek` `weekday` `weekofyear` `dayofyear` `weak` use smaller usable return type.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

### CODE REVIEW

File: be/src/vec/functions/to_time_function.cpp [hunk 1]
```diff
@@ -107,6 +121,18 @@ void register_function_to_time_function(SimpleFunctionFactory& factory) {
     factory.register_function<FunctionDateTimeV2Date>();
     factory.register_alias("date", "datev2");
     factory.register_alias("to_date", "to_datev2");
+
+    /// @TEMPORARY: for be_exec_version=2
+#define REGISTER_OLD_VERSION(FUNC, FUNC_NAME)                                      \
+    factory.register_function<FUNC##Old>(std::string {#FUNC_NAME}.append("_old")); \
```
Comment (HappenLee) [inline be/src/vec/functions/to_time_function.cpp:127]: why add "_old"

**Review by HappenLee — CHANGES_REQUESTED** [review]
why do not change the python script?

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 32.96 seconds
 stream load tsv:          424 seconds loaded 74807831229 Bytes, about 168 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          72 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230406063641_clickbench_pr_125335.html

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run clickbench

**Comment by zclllyybb** [comment]
run p0

**Comment by zclllyybb** [comment]
run clickbench

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run p0 run clickbench

**Comment by zclllyybb** [comment]
run clickbench

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/functions/function.h [hunk 2]
```diff
@@ -300,12 +302,16 @@ class IFunctionBuilder {
...
 class FunctionBuilderImpl : public IFunctionBuilder {
 public:
-    FunctionBasePtr build(const ColumnsWithTypeAndName& arguments,
-                          const DataTypePtr& return_type) const final {
+    FunctionBasePtr build(const ColumnsWithTypeAndName& arguments, const DataTypePtr& return_type,
+                          const bool skip_type_check = false) const final {
         const DataTypePtr& func_return_type = get_return_type(arguments);
-        DCHECK(return_type->equals(*func_return_type) ||
+        // check return types equal.
+        DCHECK(skip_type_check /*iff we did function replacement in exec_version 2*/ ||
```
Comment (HappenLee) [inline be/src/vec/functions/function.h:313]: which case we should skip the type check ?

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run p1

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run p0

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run p0

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**